sources/engine/Stride.VirtualReality/OculusOVR/OculusTouchController.cs

```diff
@@ -399,5 +404,31 @@ public override bool IsTouchReleased(TouchControllerButton button)
                     return false;
             }
         }
+
+        /// <summary>
+        /// Vibrate for a number of milliseconds
+        /// </summary>
+        /// <param name="duration">The number of milliseconds to vibrate for</param>
+        /// <returns></returns>
+        public override async Task Vibrate(int durationMs)
+        {
+            void SetOvrVibration(bool enable)
+            {
+                float frequency = enable ? 1 : 0;
+                OculusOvr.SetVibration(OvrSession, hand, frequency, 1);
+            }
+            ConcurrentVibratingCallCount++;
+            while (durationMs > 2000)
+            {
+                SetOvrVibration(true);
+                durationMs -= 2000;
+                await Task.Delay(2000);
+            }
+            SetOvrVibration(true);
+            await Task.Delay(durationMs);
+            ConcurrentVibratingCallCount--;
+            if (ConcurrentVibratingCallCount == 0)
+                SetOvrVibration(false);
+        }
```

Comment:
This is not thread safe as far as I can tell; if the user is calling Vibrate from the main thread while a concurrent one is finishing on a task thread this will fail since `ConcurrentVibratingCallCount` does not have any concurrent access mechanism in place.

Does the vibration automatically turn off after `2000` ms, is that why you have to periodically reset it ?

Is there a reason why frequency and amplitude are fixed values ? It sounds like those could be interesting for the caller to manipulate. For example slowly fading in and out the amplitude for an earthquake-like effect. Although those kinds of usage kind of goes against how the api is designed right now, maybe it would be best to change this into a simple `void Vibrate(int frequency, int amplitude)` and then have another non-virtual method that generates a task that wraps around this method, something like:
```cs
public async Task Vibrate(int durationMs, int frequency = 1, float amplitude = 1)
{
    Vibrate(frequency, amplitude);
    await Task.Delay(durationMs);
    StopVibration();
}
public abstract Task Vibrate(int frequency, float amplitude);
public abstract Task StopVibration();
```

Reply:
Thanks for reviewing!

You're right about the ConcurrentVibratingCallCount increment/deincrement calls not being thread safe. Would it make more sense to acquire a lock, or replace it with a queue? 

According to the docs on the ovr_SetControllerVibration called in OculusOVR.cpp, the vibration lasts for at most 2.5 seconds. 2000 milliseconds is a bit under this limit, and could be increased a bit.

According to the docs for the oculus runtime, frequency only accepts the values 0.0, 0.5, and 1.0. It clamps everything else to those values. 0.0 is no vibration, so we'd really just be adding a parameter that's either 0.5 or 1.0. There isn't a direct parallel in the openVR runtime. I was thinking we might be able to emulate it in the openVR runtime using shorter haptic pulses if we decide that the 0.5 frequency option is actually useful, but every call to TriggerHapticPulse there requires a 5ms cooldown. I didn't include the option in the PR because 50% or 100% (and no in-between) frequency option didn't seem useful, but I guess there's no downside to exposing it for oculus in case it is somehow useful to someone.

As for amplitude, the reason I didn't expose it is again because openVR doesn't seem to have a parallel, at least with the Valve.VR.OpenVR.System.TriggerHapticPulse(index, axis, duration) function. There could be another way to do it, but I wasn't able to find any. 

Does it make sense to have a vibrate with amplitude option for oculus runtime but not openVR? Apps developed specifically for the oculus store might use that feature, but games going to steam probably want to support both types of controllers. Writing this out, I'm thinking it does make sense to expose these options for the oculus runtime since there's not really a downside.